### ISSUE TITLE
Design tweaks to Function & as_function_t

### ISSUE THREAD

**Issue by CaseyCarter** [ericniebler/stl2#135]
`Function` (and its associated result type) and the `as_function_t` alias template have some shortcomings when used to constrain forwarding functions. I'm using slightly modified definitions in the implementation that I believe we should put in the TS. I'll describe the differences briefly here, and develop wording if we agree that the changes are necessary.

Algorithms accept functions and callables by value. Constraining a template that accepts an argument by forwarding reference that will eventually be forwarded to an algorithm for use as a function/callable is awkward and error-prone, requiring arcane use of `decay_t`:

```c++
template <Range R, class Pred, class Proj = identity>
  requires
    // Must decay the types of the predicate and projection:
    IndirectCallablePredicate<decay_t<Pred>, projected<iterator_t<R>, decay_t<Proj>>>()
    // ...and ensure that the decayed types are constructible from the argument types:
    && Constructible<decay_t<Pred>, Pred&&>()
    && Constructible<decay_t<Proj>, Proj&&>()
  // Is it necessary to ensure that decay_t<Pred> and decay_t<Proj> are CopyConstructible here?
  // No, but the reasoning is far from obvious to a programmer who is not intimately familiar
  // with the concept internals.
auto my_find_if(R&& range, Pred&& pred, Proj&& proj = Proj{}) {
  // ...do preliminary work...
  return ranges::find_if(std::forward<R>(range),
                         std::forward<Pred>(pred),
                         std::forward<Proj>(proj));
}
```

Constraining forwarding functions is greatly simplified if we change `Function` and `as_function_type` to perform the required decay for us, and to ensure that the decayed type is constructible from the parameter type:

```c++
template <class F, class...Args>
concept bool Function() {
  return CopyConstructible<decay_t<F>>() &&
    Constructible<decay_t<F>, F&&>() &&
    requires (decay_t<F> f, Args&&...args) {
      f(std::forward<Args>(args)...); // not required to be equality preserving
    };
}
```
It's not necessary to require `Same<decltype(f(args...)), result_of_t<F&(Args...)>>()`, since they are by definition equivalent. We also don't want to couple the concept to the design of `result_of_t`, which necessarily tracks `std::invoke`. Adjusting `Function` in this way necessitates a bit of smarts in the associated result type as well to account for the decay:

```c++
Function{F, ...Args}
using result_t = decltype(declval<decay_t<F>&>()(declval<Args>()...));
```

Note that `result_t` is constrained to require that `Function<F, Args...>()` be satisfied: the implementation constrains associated types to get better and "early" errors.

Modifications to `as_function_t` are similar:

```c++
// exposition only
template <class F>
struct __as_callable_wrapper {
  using type = ext::callable_wrapper<F>;
};

template <class F>
  requires CopyConstructible<decay_t<F>>() &&
    Constructible<decay_t<F>, F&&>()
using as_function_t = meta::_t<meta::if_<
  is_member_pointer<decay_t<F>>, __as_callable_wrapper<decay_t<F>>, decay<F>>>;
```
(Note: `callable_wrapper<F>` is a call wrapper type that forwards to `invoke` which I am not proposing.)

I also propose that `as_function_t` is overspecified in the TS; it requires exactly `Same<as_function_t<F>, decltype(mem_fn(declval<F>))>()` when `F` is a member pointer type, and `Same<as_function_t<F>, decay_t<F>>()` otherwise. Since `as_function_t` is simply a tool to constrain declarations, it is sufficient to require that (a) `Function<as_function_t<F>, Args...>()` be satisfied if and only if `INVOKE(declval<decay_t<F>&>(), declval<Args>()...)` is valid, and (b) that `result_t<as_function_t<F>, Args...>` be the same type as `result_of_t<add_lvalue_reference<decay_t<F>>(Args...)>`. This allows implementor freedom (e.g., my implementation uses the aforementioned `callable_wrapper` in the algorithms, so specifying it in `as_function_t` amortizes the template instantiation cost).

Since these changes to `Function`, `result_t` and `as_function_t` affect all of the related `Function` and `Callable` concepts, the end result is to greatly simplify the example forwarding function definition to:

```c++
template <Range R, class Pred, class Proj = identity>
  requires IndirectCallablePredicate<Pred, projected<iterator_t<R>, Proj>>()
auto my_find_if(R&& range, Pred&& pred, Proj&& proj = Proj{}) {
  // ...do preliminary work...
  return ranges::find_if(std::forward<R>(range),
                         std::forward<Pred>(pred),
                         std::forward<Proj>(proj));
}
```

which works even with ridiculous function object types:

```c++
struct foo {
  foo(foo&&) = default;
  foo(const foo&) = default;
  foo(volatile foo&&) {}

  int operator()(int i) &;
};

template <class...>
constexpr bool test() { return false; }

template <ranges::Function<int>...Fs>
  requires (ranges::Same<int, ranges::result_t<Fs, int>>() && ...)
constexpr bool test() { return true; }

static_assert(test<foo, foo&, foo&&, foo const, foo const&,
                   foo const&&, foo volatile, foo volatile&&>());
```

**Comment by CaseyCarter** [comment]
I suppose I should also not that I the use of `decay_t` in the function concepts allows them to be satisfied by function types in addition to function *object* types: `Function<int(int, double), int, double)>()` for example. I don't use that feature for anything, but I imagine it could be convenient for metaprogramming.

**Comment by CaseyCarter** [comment]
I'm still acclimating to snake_case associated types. The name `result_t` is clearly so similar to `result_of_t` as to cause confusion and needs bikeshedding. Perhaps `function_result_t`?

**Comment by ericniebler** [comment]
> The name  result_t  is clearly so similar to  result_of_t  as to cause confusion and needs bikeshedding.

I currently am using the existing `result_of_t` directly. I think that's the right answer.

**Comment by CaseyCarter** [comment]
> I currently am using the existing result_of_t directly. I think that's the right answer.

The issues I have with `result_of_t` are (1) the hazards of using it in a forwarding function, where you actually want `result_of_t<decay_t<F>&, Args...>` since the final callsite inside some template that accepts the expression by value will be calling a decayed lvalue, and (2) we cannot possibly constrain `result_of_t` to require `Function` because of breakage. I've really enjoyed the results of pushing constraints out into the "leaves" of the implementation, it results in better/earlier errors and makes specification of constraints and metaprogramming easier. Losing that for `ResultType` doesn't break the design but it is a shame.

> Shall we fold `Callable` into `Function`? The Ranges TS doesn't actually use the function call syntax in many places, and `INVOKE` can be safely used everywhere.

I was thinking about this last night after writing this post. We could easily steal `std::invoke` from the WP much as we have `std::size` - the specification is only a couple of sentences "blah blah evaluates INVOKE(...)". The plain `Function` concepts are really only used to specify the `Callable` concepts: indirectly through the `Indirect` concepts in the TS, and literally in cmcstl2 which has `Callable`, `RegularCallable`, `CallablePredicate` and `CallableRelation`. I know of [at least one freely available implementation of `invoke` using concepts](https://github.com/CaseyCarter/cmcstl2/blob/master/include/stl2/detail/functional/invoke.hpp) so it wouldn't be a burden for implementors.

`Function` family has always felt pretty STL-specific to me, since it's a composition of `callable function object` and `CopyConstructible`, I think if we aren't providing the decomposed flavors of those two notions that it wouldn't hurt many use cases to change to `Callable`.

**Comment by ericniebler** [comment]
My preference is to drop `Function` in favor of `Callable`, provide `std::invoke`, use `result_of_t` directly, and don't do the `decay_t` dance in the concept. I honestly don't see a need for it. Callables aren't perfectly forwarded in the STL and they arguably should never be. The model is and always has been to use `std::ref` when passing a stateful callable to an algorithm ... which is exceptionally rare in my experience.

**Comment by CaseyCarter** [comment]
I believe this issue is effectively closed by b5dc238bfb03aa934d9d1a69e26bdb18381e69d7.
